Added rudimentary item support.

coreknee opened the pull request:
This PR adds rudimentary item support to the engine, It also makes the menuless sample fight give the same amount of hp dealt to the Omega Dummy.

coreknee committed:
exist

coreknee committed:
get rid of heal button

was left over from debugging

bravelycowering commented:
I'm probably not going to merge this until it's more complete to avoid adding even more unfinished code to the engine, but it's a good start.

coreknee committed:
e

coreknee committed:
uh

coreknee committed:
added a sample battle for the items, gave the soul four more default items

coreknee commented:
Ive added some new functions related to the items, this also comes with the ability to make items that "dissapear" (not really but in my defense, technical limits.) when used, It also allows for the player to gain items. (Possibly useful for making a ACT like Dream from the Asriel fight which gives you a item, it does require you to keep track of which items are empty items and which are actual items however i believe it works for the purpose.), There is also now a example script of a battle with no items.

bravelycowering commented:
Looks good, though a few things could be cleaned up before merging:
- You can remove an item from a table by doing `table.remove(tbl, index)` rather than replacing an item with N/A and making it do nothing on use
- A few wrapper functions for making items could be useful, stuff for implementing food, weapons, and armor easily that way making multiple healing items for a fight is easier
- The item menu should function similarly to Undertale's, with every 4 items wrapping to a new page, though this probably requires some new code in `dialogue.lua` to be written

If I find the time to do so, I'll try to work on some of these so there's less to clean up before items are feature complete.

On another note, I found an old implementation of items I was working on a while back on my laptop. but I never pushed it to the GitHub page because there were a few technical issues (I don't remember what they were, it seems to work fine but I don't trust it and don't have time to extensively test them). I've uploaded those changes under the old-item-testing branch if you want to mess around with it.

coreknee commented:
I didnt do table.remove(tbl,index) due to the issue of the index the item uses to get rid of itself being inaccurate if a item before it is removed, For example:
Item 2 has the index as 2,
Item 1 gets removed from the table, making Item 2 Item 1.
Item 2 still has the same index, thefore when eaten the next item will either be eaten or the game will crash (if there is no other item)

coreknee committed:
Merge branch 'main' into main

coreknee committed:
remove dead code, make makehealingitem function

coreknee committed:
Merge branch 'main' of https://github.com/emir4169/undertale-ceiling-fan

bravelycowering commented:
If you have the battle script remove the item when used based on index that shouldn't be an issue. This implementation is in the `old-item-testing` branch I mentioned earlier.

bravelycowering commented:
I went ahead and added multi page menus, which can be created by adding a 4th argument to `dialogue:makechoices` for the maximum amount of rows per page.

coreknee committed:
id system so issues wont arise

coreknee committed:
Merge branch 'main' of https://github.com/emir4169/undertale-ceiling-fan

coreknee committed:
git hell i am in

coreknee committed:
escape from git hell 1

coreknee committed:
the escape from git hell 2

coreknee commented:
wait did i accidentally close it woopğs

coreknee committed:
the final escape from git hell

coreknee commented:
> I went ahead and added multi page menus, which can be created by adding a 4th argument to `dialogue:makechoices` for the maximum amount of rows per page.

I have implemented this to the items implementation. I have also made a ID system for each item such that the correct item gets removed from the array. After testing i believe that for this to be truly complete i would simply have to add armour and weapon items.

bravelycowering commented:
Seems to work great! I'll probably wind up moving item definition to the battle script in the future and define some default behaviors for items like equipping them to armor and defense. I'll review this more in depth when I have the time and merge it, thanks!

coreknee committed:
makes sure we dont get a annoying warning about how newText is deprecated in love 12 by making a function that works as a wrapper

coreknee commented:
reminder to review!

bravelycowering commented:
Ah sorry, I've been a bit preoccupied, I'll get back to this at some point.

coreknee commented:
reminder to review!

coreknee commented:
please review

coreknee commented:
reminder to review

coreknee commented:
hello i am asking for a review

coreknee committed:
fix mods folder not being correctly opened on linux

coreknee committed:
framerate issue fixer from 2011, make soul code account for the 30fps env, tidy up soul code a bit

coreknee committed:
adjust fightbar object for 30fps

coreknee committed:
change stuff for 30fps

coreknee committed:
how did i not notice this one

coreknee committed:
fix minor bug

coreknee committed:
merge PR #3

bravelycowering commented:
Hello, I'm going to close this PR as it not only now contains code from other PRs and unrelated code to the items system, but I've already got an items system in another branch. If you'd like to open other PRs for the various bug fixes individually feel free
